src/Storages/MergeTree/MergeTreeDataPartWriterCompact.cpp

```diff
@@ -163,6 +166,49 @@ void MergeTreeDataPartWriterCompact::finishDataSerialization(IMergeTreeDataPart:
     stream.reset();
 }
 
+void fillIndexGranularityImpl(
+    MergeTreeIndexGranularity & index_granularity,
+    size_t & index_offset,
+    size_t index_granularity_for_block,
+    size_t rows_in_block)
+{
+    bool initial_marks_count = index_granularity.getMarksCount();
+    for (size_t current_row = index_offset; current_row < rows_in_block; current_row += index_granularity_for_block)
+    {
+        size_t rows_left_in_block = rows_in_block - current_row;
+
+        /// Try to extend last granule if block is large enough
+        ///  or it isn't first in granule (index_offset != 0).
+        if (rows_left_in_block < index_granularity_for_block &&
+            (rows_in_block >= index_granularity_for_block || index_offset != 0))
+        {
+            // If enough rows are left, create a new granule. Otherwise, extend previous granule.
+            // So, real size of granule differs from index_granularity_for_block not more than 50%.
+            if (rows_left_in_block * 2 >= index_granularity_for_block)
+                index_granularity.appendMark(rows_left_in_block);
+            else
+                index_granularity.addRowsToLastMark(rows_left_in_block);
+        }
+        else
+        {
+            index_granularity.appendMark(index_granularity_for_block);
+        }
+    }
+
+    /// Primary index shouldn't be written, if there was no granule added
+    if (initial_marks_count == index_granularity.getMarksCount())
+        index_offset = std::max(index_offset, rows_in_block);
```

Comment:
I don't understand why it helps :( Very complicated logic of `index_offset` update.

Reply:
It helps, because [here](https://github.com/ClickHouse/ClickHouse/blob/d952e5aba22586f94db89705e452a8266506c209/src/Storages/MergeTree/IMergeTreeDataPartWriter.cpp#L228) we won't run loop. But I agree, that it's became too complicated. I'll try to rewrite this part of code.